docs/mocksupport.rst

```diff
@@ -867,13 +738,338 @@ This adds the instances to the repository display of the previous example:
+generate responses that have side effects, or manage the returns differently
+than the normal responses.  Thus, for example, the DMTF defined CIM_Namespace
+class returns instances of each of the namespaces in the CIM repository. That means
+that somewhere an instance the class must be created each time a namespace is
+created or the responder must get the set of namespaces from the CIM repository
+to generate responses to ``GetInstance``, ``EnumerateInstance``, etc.
+
+Also, the :meth:`pywbem_mock.FakedWBEMConnection.InvokeMethod` client operation
+depends on a specific responder method in the WBEM server since the behavior of
+WBEM server methods is dependent on the specific method.
+
+To meet these needs, pywbem_mock has implemented  the capability to create
+user-defined providers :term:`user-defined provider` for selected classes and
+operations which can can manipulate the client input and generate responses
+specific for defined CIM classes.
+
+User-defined providers can be writted by the user by defining subclasses of
+specific provider classes used by the fake WBEM server and registering these
+classes as providers using
+:meth:`~pywbem_mock.FakedWBEMConnection.register_provider`.
+
+Pywbem defines user-defined provider types so that specific request operations
+can be executed for each provider type.
+
+
+.. table:: Provider types in pywbem_mock
+
+    +-----------------------+-------------+------------------------+-----------------------+
+    | Provider Type         +  type name  + CIM Request Operations | Superclass            |
+    +=======================+======================================+=======================+
+    |  method provider      | "method"    | InvokeMethod           | MethodProvider        |
+    |  instance write       | "instance"  | CreateInstance         | InstanceWriteProvider |
+    |  instance write       | "instance"  | ModifyInstance         | InstanceWriteProvider |
+    |  instance write       | "instance"  | DeleteInstance         | InstanceWriteProvider |
+    +-----------------------+-------------+------------------------+-----------------------+
+
+
+Each user-defined provider consists of a python class which must be a subclass
+of the corresponding standard provider for the provider type.
+
+.. code-block:: python
+
+    # define a subclass to to handle InvokeMethod from the MethodProvider
+    class MyUserProvider(MethodProvider):
+
+Within that class, the following components must be defined.
+
+1. Constructor (`__init__()`) that takes as input at least the ``cimrepository``
+   object parameter and passes it to the superclass and any other constructor
+   parameters the user-defined provider requires.  Since the user creates the
+   instance.
+
+.. code-block:: python
+
+    def __init__(self, cimrepository):
+        """
+        Init of test provider
+        """
+        super(UserMethodTestProvider, self).__init__(cimrepository)
```

Comment:
I suggest that this piece of code be consistent with the class header a few lines up, i.e. use the same class name.

Reply:
DONE